pkg/apis/autoscaling/v1alpha1/metricsproducer_validation.go

```diff
@@ -12,40 +12,61 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// +kubebuilder:webhook:verbs=create;update,path=/validate-autoscaling-karpenter-sh-v1alpha1-metricsproducer,mutating=false,sideEffects=None,failurePolicy=fail,groups=autoscaling.karpenter.sh,resources=metricsproducers,versions=v1alpha1,name=vmetricsproducer.kb.io
 package v1alpha1
 
-import "fmt"
+import (
+	"fmt"
+	"k8s.io/apimachinery/pkg/runtime"
+	"reflect"
+	"regexp"
+	"strings"
+	"time"
+)
+type specValidator interface {
+	validate() error
+}
 
-// +kubebuilder:object:generate=false
-type QueueValidator func(*QueueSpec) error
+// Only Validates ScheduledCapacity MetricsProducer right now
+func (m *MetricsProducer) ValidateCreate() error {
+	v := reflect.ValueOf(m)
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+		if !field.IsNil() {
+			return field.Interface().(specValidator).validate()
+		}
+	}
+	return nil
+}
 
-var (
-	queueValidator = map[QueueType]QueueValidator{}
-)
+func (m *MetricsProducer) ValidateUpdate(old runtime.Object) error {
```

Comment:
If our validation logic is all the same, let's do something like
```
func (m *MetricsProducer) ValidateCreate() error {
  return m.validate()
}
func (m *MetricsProducer) ValidateUpdate() error {
  return m.validate()
}

func (m *MetricsProducer) validate() error {
 ... 
}
```
This is a bit more idiomatic, since it makes "create" less special.